### PR TITLE
Attaching legs to a snake

### PR DESCRIPTION
**Pull request by kurazu**
Presentation *Attaching legs to a snake - or - Python3 extensions* by *Tomasz Maćkowiak*

### CODE REVIEW

**Comment by nihilifer** [comment]
LGTM

File: 2015/presentations/attaching_legs_to_a_snake/text.md [hunk 1]
```diff
@@ -0,0 +1,582 @@
+# Attaching legs to a snake - or - Python3 extensions - Tomasz Maćkowiak
+
+## Introduction
+
+Python is a powerful language by itself. The robust standard library and the myriad of additional packages make it a Swiss Army knife of programming languages. But sometimes it is not enough. When you need to integrate with a low-level C API or execute heavy computations without the interpreter overhead, writing a Python extension module might be a good idea.
```
Comment (soltysh) [inline 2015/presentations/attaching_legs_to_a_snake/text.md]: The robust standard library and the myriad of additional packages make it a Swiss Army knife for any programming task.

File: 2015/presentations/attaching_legs_to_a_snake/text.md [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+
+This article will show you how to write, build and run an extension; what C API you can use in it and what to remember about; how to link to shared libraries and what does Boost.Python simplify.
```
Comment (soltysh) [inline 2015/presentations/attaching_legs_to_a_snake/text.md]: what to remember ~~about~~

File: 2015/presentations/attaching_legs_to_a_snake/text.md [hunk 3]
```diff
@@ -0,0 +1,582 @@
...
+   "basic",   /* name of module */
+   "The simplest module", /* module documentation, may be NULL */
+   -1,       /* size of per-interpreter state of the module,
+                or -1 if the module keeps state in global variables. */
+   basic_methods
+};
+```
+ We pass its name, docstring, size of the optional module state memory block and function definitions.
+
+### Hearth
```
Comment (soltysh) [inline 2015/presentations/attaching_legs_to_a_snake/text.md]: Heart

File: 2015/presentations/attaching_legs_to_a_snake/text.md [hunk 4]
```diff
@@ -0,0 +1,582 @@
...
+The most important, though, is the module initialisation function that creates the module on the basis of the module definition:
+```
+PyMODINIT_FUNC
+PyInit_basic(void)
+{
+    return PyModule_Create(&basic_module);
+}
+```
+
+## Animation - how to make it come alive
```
Comment (soltysh) [inline 2015/presentations/attaching_legs_to_a_snake/text.md]: how to make it alive

File: 2015/presentations/attaching_legs_to_a_snake/text.md [hunk 5]
```diff
@@ -0,0 +1,582 @@
...
+
+## Digestive system - parsing parameters
+
+### Positional parameters
+
+To parse positional parameters we need to define our function with ```METH_VARARGS``` flag in the module's functions definition:
+```
+{"hello",  param_hello, METH_VARARGS, "Say hello."},
+```
+With such a declaration our function will get one more parameter, similar to ```*args``` construct in pure Python:
```
Comment (soltysh) [inline 2015/presentations/attaching_legs_to_a_snake/text.md]: With such a declaration, our

File: 2015/presentations/attaching_legs_to_a_snake/text.md [hunk 6]
```diff
@@ -0,0 +1,582 @@
...
+To get an item under a given key in a dictionary (```category_sequence = mapping[category]```), we use:
+```
+PyObject * category_sequence = PyObject_GetItem(mapping, category);
+if (category_sequence == NULL) {
+    return NULL;
+}
+// Deal with the object ...
+```
+
+To check if a sequence contains a given item (```contains = item in category_sequence```), we use:
```
Comment (soltysh) [inline 2015/presentations/attaching_legs_to_a_snake/text.md]: ~~a~~ given

File: 2015/presentations/attaching_legs_to_a_snake/text.md [hunk 7]
```diff
@@ -0,0 +1,582 @@
...
+static PyMemberDef Native_members[] = {
+    {"name", T_OBJECT_EX, offsetof(Native, name), 0, "Name"},
+    {"number", T_LONG, offsetof(Native, number), 0, "Number"},
+    {"pointer", T_STRING, offsetof(Native, pointer), READONLY, "Pointer"},
+    {NULL}  /* Sentinel */
+};
+```
+For each member we define a name, type, offset in class structure, flags and a docstring.
+
+We can also define what methods will be available on our objects:
```
Comment (soltysh) [inline 2015/presentations/attaching_legs_to_a_snake/text.md]: s/what/which

File: 2015/presentations/attaching_legs_to_a_snake/text.md [hunk 8]
```diff
@@ -0,0 +1,582 @@
...
+{
+    def("has_letter", has_letter);
+}
+```
+
+Extensions written using *Boost.Python* can be much more concise.
+
+### Compiling and linking
+
+*Boost.Python* comes as a shared library. That means that during compilation and executing of our extension Python needs to be able to read the library's files. If the library is installed system-wide, we don't have to worry about paths. If the library is installed locally, we need to remember to pass the correct include path and library path during compilation and to have the ```LD_LIBRARY_PATH``` system variable correctly set during running.
```
Comment (soltysh) [inline 2015/presentations/attaching_legs_to_a_snake/text.md]: of our extension, Python...

**Comment by soltysh** [comment]
Small nits, I'll address after mergeing, LGTM.

**Comment by soltysh** [comment]
Thanks Tomasz!

**Pull request merged**